### PR TITLE
[Feature] Improve oracle spec

### PR DESCRIPTION
**Pull request by YunSuk-Yeo**
## Summary of changes

* Append aggregate msgs
* Change oracle param `Whitelist` struct to include `TobinTax` 
* Keep tracking `VoteTargets` in the Keeper

This will enable the voters to do oracle multiple votes with an aggregate msg which is highly downsized compared with normal prevote/vote. 

The param structure change will enable to enter illiquid factor, which determines the swap spread, at whitelist registration step via gov proposal.

The core directly refer `params.Whitelist` to check oracle slashing at columbus-3. If a new denom is appended to `params.Whitelist`, then it can cause some validators miss a oracle vote because it comes middle of a oracle vote period. In order to prevent this confusion, the oracle keeper will keep `VoteTargets` and update it at the end of vote period with `params.Whitelist`.

See discussion: [Agora link](https://agora.terra.money/t/oracle-optimization-specification/211)

## Report of required housekeeping

- [x] Github issue OR spec proposal link
- [x] Wrote tests
- [ ] Updated relevant documentation (docs/)
- [ ] Added a relevant changelog entry: clog add [section] [stanza] [message]

----

## (FOR ADMIN) Before merging

- [x] Added appropriate labels to PR
- [x] Squashed all commits, uses message "Merge pull request #XYZ: [title]" (coding standards)
- [x] Confirm added tests are consistent with the intended behavior of changes
- [x] Ensure all tests pass

### CODE REVIEW

File: x/oracle/client/cli/tx.go [hunk 1]
```diff
@@ -208,3 +204,122 @@ where "terra1..." is the address you want to delegate your voting rights to.
...
+// GetCmdAggregateExchangeRateVote will create a aggregateExchangeRateVote tx and sign it with the given key.
+func GetCmdAggregateExchangeRateVote(cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "aggregate-vote [salt] [exchange_rates] [validator]",
+		Args:  cobra.RangeArgs(2, 3),
+		Short: "Submit an oracle vote for the exchange_rates of Luna",
+		Long: strings.TrimSpace(`
+Submit a vote for the exchange_rates of Luna w.r.t the input denom. Companion to a prevote submitted in the previous vote period. 
+
+$ terracli tx oracle vote 1234 8888.0ukrw,1.243uusd,0.99usdr
```
Comment (hanjukim) [inline x/oracle/client/cli/tx.go]: isn't vote supposed to be aggregate-vote here?

**Comment by dokwon** [comment]
@YunSuk-Yeo instead of `VoteTarget`, how do you feel about calling these `WhitelistWaitQueue`? Seems more appropriate given the data structure stores denominations that are about to be entered into the whitelist. Do we need to take the same precaution for removing denoms from the whitelist? (assuming this is also an option)

File: app/sim_test.go [hunk 2]
```diff
@@ -399,7 +399,7 @@ func testAndRunTxs(app *TerraApp) []simulation.WeightedOperation {
 		{
 			func(_ *rand.Rand) int {
 				var v int
-				ap.GetOrGenerate(cdc, OpWeightMsgUnjail, &v, nil,
+				ap.GetOrGenerate(cdc, OpWeightMsgSend, &v, nil,
```
Comment (dokwon) [inline app/sim_test.go]: Note: not relevant to oracle improvement, tag along to the PR

File: client/lcd/swagger-ui/swagger.yaml [hunk 3]
```diff
@@ -2408,6 +2408,100 @@ paths:
           description: Bad Request
         500:
           description: Internal Server Error
+  /oracle/voters/{validator}/aggregate_prevote:
+    post:
+      summary: Generate oracle aggregate exchange rate prevote message containing hash
```
Comment (dokwon) [inline client/lcd/swagger-ui/swagger.yaml]: containing *a* hash

File: client/lcd/swagger-ui/swagger.yaml [hunk 4]
```diff
@@ -2408,6 +2408,100 @@ paths:
...
+          description: OK
+          schema:
+            $ref: "#/definitions/AggregateExchangeRatePrevote"
+        400:
+          description: Bad Request
+        500:
+          description: Internal Server Error
+  /oracle/voters/{validator}/aggregate_vote:
+    post:
+      summary: Generate oracle aggregate exchange rate vote message containing exchange rates and salt to prove aggregate prevote
```
Comment (dokwon) [inline client/lcd/swagger-ui/swagger.yaml]: to prove *the* aggregate prevote

File: contrib/devtools/install-golangci-lint.sh [hunk 5]
```diff
@@ -21,7 +21,7 @@ get_latest_release() {
     grep '"tag_name":' | sed -E 's/.*"([^"]+)".*/\1/'
 }
 
-VERSION="$(get_latest_release golangci/golangci-lint)"
+# VERSION="$(get_latest_release golangci/golangci-lint)"
```
Comment (dokwon) [inline contrib/devtools/install-golangci-lint.sh]: why is this commented out? @YunSuk-Yeo

File: contrib/devtools/install-golangci-lint.sh [hunk 6]
```diff
@@ -31,4 +31,4 @@ echo "Checking hashsum ..." >&2
 chmod +x "${installer}"
 
 echo "Launching installer ..." >&2
-exec "${installer}" -d -b "${GOBIN}" "${VERSION}"
\ No newline at end of file
+exec "${installer}" -d -b "${GOBIN}" "${VERSION}"
```
Comment (dokwon) [inline contrib/devtools/install-golangci-lint.sh]: Isn't this the same line?

File: x/market/internal/keeper/swap.go [hunk 7]
```diff
@@ -70,18 +70,22 @@ func (k Keeper) ComputeSwap(ctx sdk.Context, offerCoin sdk.Coin, askDenom string
...
+		if err2 != nil {
+			return sdk.DecCoin{}, sdk.Dec{}, err2
+		}
+
+		// Apply highest illiquid factor for the denoms in the swap operation
+		if askIlliquidFactor.GT(illiquidFactor) {
+			illiquidFactor = askIlliquidFactor
+		}
+
+		spread = spread.Mul(illiquidFactor)
```
Comment (dokwon) [inline x/market/internal/keeper/swap.go]: Think it should be better data structure design to have unique tobin taxes registered for each of the whitelisted denoms instead of having a universal tobin tax and an "illiquid factor" for each denom.

File: x/market/internal/types/params.go [hunk 8]
```diff
@@ -44,21 +42,19 @@ var _ subspace.ParamSet = &Params{}
...
 func DefaultParams() Params {
 	return Params{
-		BasePool:             DefaultBasePool,
-		PoolRecoveryPeriod:   DefaultPoolRecoveryPeriod,
-		MinSpread:            DefaultMinSpread,
-		TobinTax:             DefaultTobinTax,
-		IlliquidTobinTaxList: DefaultIlliquidTobinTaxList,
+		BasePool:           DefaultBasePool,
+		PoolRecoveryPeriod: DefaultPoolRecoveryPeriod,
+		MinSpread:          DefaultMinSpread,
```
Comment (dokwon) [inline x/market/internal/types/params.go]: What do you think about renaming `MinSpread` to `MinStabilitySpread`? Given that the `MinSpread` only applies to swaps involving Luna, think we need to differentiate this variable name somehow to remind people of that restriction

File: x/oracle/internal/types/msgs.go [hunk 9]
```diff
@@ -212,3 +212,144 @@ func (msg MsgDelegateFeedConsent) String() string {
...
+}
+
+// GetSigners implements sdk.Msg
+func (msg MsgAggregateExchangeRatePrevote) GetSigners() []sdk.AccAddress {
+	return []sdk.AccAddress{msg.Feeder}
+}
+
+// ValidateBasic Implements sdk.Msg
+func (msg MsgAggregateExchangeRatePrevote) ValidateBasic() sdk.Error {
+
```
Comment (dokwon) [inline x/oracle/internal/types/msgs.go]: Shouldn't you validate the exchangeRates string here? e.g. check for correct comma separation, etc

File: x/oracle/internal/types/vote_test.go [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+	_, err = ParseExchangeRateTuples(duplicatedDenom)
+	require.Error(t, err)
+
+	invalidCoin := "123.123"
+	_, err = ParseExchangeRateTuples(invalidCoin)
+	require.Error(t, err)
+
+	invalidCoinWithValid := "123.0uluna,123.1"
+	_, err = ParseExchangeRateTuples(invalidCoinWithValid)
+	require.Error(t, err)
```
Comment (dokwon) [inline x/oracle/internal/types/vote_test.go]: Might want to add a test for empty string, as well as a string containing the denom but with excess characters e.g. "123ukrww", as well as a test containing a strange denom e.g. "123ajdsfjfaddaj"

File: contrib/devtools/install-golangci-lint.sh (hunk 6 above)
Comment (YunSuk-Yeo) [reply contrib/devtools/install-golangci-lint.sh]: New line added (my code editor enforcing this now).
It already covered in other PR so I will revert this file

File: x/market/internal/types/params.go (hunk 8 above)
Comment (YunSuk-Yeo) [reply x/market/internal/types/params.go]: agree,

File: x/oracle/internal/types/msgs.go (hunk 9 above)
Comment (YunSuk-Yeo) [reply x/oracle/internal/types/msgs.go]: In prevote, we do not receive exchangeRates string. We are doing this in vote ValidateBasic() with `ParseExchangeRateTuples` function

File: x/oracle/internal/types/vote_test.go (hunk 10 above)
Comment (YunSuk-Yeo) [reply x/oracle/internal/types/vote_test.go]: msg validation has no context, so it is hard to test denom is in whitelist or not.
We can do whitelist verification in handler now with voting targets.

Comment (dokwon) [reply x/oracle/internal/types/vote_test.go]: Makes sense

File: x/oracle/internal/types/msgs.go (hunk 9 above)
Comment (dokwon) [reply x/oracle/internal/types/msgs.go]: Makes sense

File: x/market/internal/types/params.go (hunk 8 above)
Comment (dokwon) [reply x/market/internal/types/params.go]: Awesome

File: x/market/internal/keeper/params.go [hunk 11]
```diff
@@ -17,9 +17,9 @@ func (k Keeper) BasePool(ctx sdk.Context) (res sdk.Dec) {
 	return
 }
 
-// MinSpread is the minimum swap fee(spread)
-func (k Keeper) MinSpread(ctx sdk.Context) (res sdk.Dec) {
-	k.paramSpace.Get(ctx, types.ParamStoreKeyMinSpread, &res)
+// MinStabilitySpread is the minimum swap fee(spread)
```
Comment (dokwon) [inline x/market/internal/keeper/params.go]: Would add a comment that MinStabilitySpread only applies to swaps involving Luna 

--> // MinStabilitySpread is the minimum spread applied to swaps to / from Luna. Intended to prevent swing trades exploiting oracle period delays

File: x/oracle/internal/keeper/vote_target.go [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+// IsVoteTarget returns existence of a denom in the the voting target list
+func (k Keeper) IsVoteTarget(ctx sdk.Context, denom string) bool {
+	_, err := k.GetTobinTax(ctx, denom)
+	return err == nil
+}
+
+// GetVoteTargets returns the voting target list on current vote period
```
Comment (dokwon) [inline x/oracle/internal/keeper/vote_target.go]: // GetVoteTargets returns the voting target denom list on current vote period

File: x/oracle/internal/types/keys.go [hunk 13]
```diff
@@ -84,13 +83,13 @@ func GetAggregateExchangeRateVoteKey(v sdk.ValAddress) []byte {
...
-func GetIlliquidFactorKey(d string) []byte {
-	return append(IlliquidFactoerKey, []byte(d)...)
+// GetTobinTaxKey - stored by *denom* bytes
+func GetTobinTaxKey(d string) []byte {
+	return append(TobinTaxKey, []byte(d)...)
 }
 
-// SplitDenomFromIlliquidFactorKey - split denom from the illiquid factor key
-func SplitDenomFromIlliquidFactorKey(key []byte) (denom string) {
+// SplitDenomFromTobinTaxKey - split denom from the tobin tax key
```
Comment (dokwon) [inline x/oracle/internal/types/keys.go]: SplitDenomFromTobinTaxKey --> ExtractDenomFromTobinTaxKey

File: x/oracle/internal/keeper/vote_target.go (hunk 12 above)
Comment (dokwon) [reply x/oracle/internal/keeper/vote_target.go]: Resolved

File: x/market/internal/keeper/swap.go (hunk 7 above)
Comment (dokwon) [reply x/market/internal/keeper/swap.go]: Resolved

File: client/lcd/swagger-ui/swagger.yaml (hunk 4 above)
Comment (dokwon) [reply client/lcd/swagger-ui/swagger.yaml]: resolved

File: client/lcd/swagger-ui/swagger.yaml (hunk 3 above)
Comment (dokwon) [reply client/lcd/swagger-ui/swagger.yaml]: resolved

File: contrib/devtools/install-golangci-lint.sh (hunk 5 above)
Comment (dokwon) [reply contrib/devtools/install-golangci-lint.sh]: resolved

File: x/oracle/internal/types/keys.go (hunk 13 above)
Comment (dokwon) [reply x/oracle/internal/types/keys.go]: Resolved

File: x/oracle/client/cli/tx.go (hunk 1 above)
Comment (YunSuk-Yeo) [reply x/oracle/client/cli/tx.go]: right

Comment (dokwon) [reply x/oracle/client/cli/tx.go]: Resolved
